src/System.Net.Http/src/System/Net/Http/Unix/CurlCallbacks.cs

```diff
@@ -0,0 +1,240 @@
+                    if (string.IsNullOrWhiteSpace(responseHeader))
+                    {
+                        state.ResponseMessage.SignalHeadersReceived();
+                        return totalSize;
+                    }
+
+                    // TODO: Understand scenarios where multiple sets of headers are received
+                    if (!TryParseStatusLine(response, responseHeader))
+                    {
+                        int colonIndex = responseHeader.IndexOf(':');
+
+                        // Skip malformed header lines that are missing the colon character.
+                        if (colonIndex > 0)
+                        {
+                            string headerName = responseHeader.Substring(0, colonIndex);
+                            string headerValue = responseHeader.Substring(colonIndex + 1);
+
+                            if (!response.Headers.TryAddWithoutValidation(headerName, headerValue))
+                            {
+                                // TODO: Skip compression related headers
+                                response.Content.Headers.TryAddWithoutValidation(headerName, headerValue);
+                            }
+                        }
+                    }
+                    return totalSize;
+                });
+        }
+
+        private ulong CurlReceiveBodyCallback(
+            IntPtr buffer,
+            ulong size,
+            ulong nitems,
+            IntPtr context)
+        {
+            return ExecuteCallback(buffer, size, nitems, context,
+                (ptr, totalSize, state) =>
+                {
+#if DEBUG
+                    LogCurlBuffer(ptr, totalSize);
+#endif
+
+                    // TODO: Handle compression and chunked-mode
+
+                    var byteBuffer = new byte[totalSize];
+                    Marshal.Copy(ptr, byteBuffer, 0, byteBuffer.Length);
+                    Stream contentStream = state.ResponseMessage.ContentStream;
+                    contentStream.Write(byteBuffer, 0, byteBuffer.Length);  // will block if no readers
+                    contentStream.Flush();
+                    return totalSize;
+                });
+        }
+
+
+        private ulong CurlSendCallback(
+            IntPtr buffer,
+            ulong size,
+            ulong nitems,
+            IntPtr context)
+        {
+            return ExecuteCallback(buffer, size, nitems, context,
+                (ptr, totalSize, state) =>
+                {
+                    state.CancellationToken.ThrowIfCancellationRequested();
+                    Stream contentStream = state.RequestContentStream;
+
+                    // TODO: Handle chunked-mode
+
+                    int bytesRead = 0;
+                    while (bytesRead < (int)totalSize)
+                    {
+                        state.CancellationToken.ThrowIfCancellationRequested();
+                        var byteBuffer = new byte[totalSize];
```

Comment:
If pooling is done as suggested earlier, than this buffer could be used in that pool as well.  At the very least, this byte[] should be allocated outside of this loop, not on every iteration.

(For both this and the response case, if we had support on Stream for working with byte* directory rather than byte[], we wouldn't need these intermediate buffers. cc: @KrzysztofCwalina)